knimin/lib/data_access.py

```diff
@@ -2759,6 +2767,155 @@ def delete_samples(self, ids):
                          RETURNING sample_id"""
                 TRN.add(sql, [sample_id])
 
+    def _study_sample_are_associated(self, study_id, sample_id):
+        """Checks whether a study and a sample are associated
+
+        Parameters
+        ----------
+        study_id : str
+            ID of the study to check
+        sample_id : str
+            ID of the sample to check
+
+        Returns
+        ------
+        bool
+            Whether they are associated
+        """
+        sql = """SELECT EXISTS (SELECT 1 FROM pm.study_sample
+                                WHERE study_id = %s
+                                AND sample_id = %s)"""
+        return self._con.execute_fetchone(sql, [study_id, sample_id])[0]
+
+    def associate_samples_with_study(self, study_id, sample_ids):
+        """Creates associations between one or more samples and one study
+
+        Parameters
+        ----------
+        study_id : int
+            ID of the study
+        sample_ids : list of str
+            IDs of the samples
+
+        Raises
+        ------
+        ValueError
+            If the study ID does not exist
+            If the study and the sample are already associated
+        """
+        self._study_exists(study_id)
+        with TRN:
+            for sample_id in sample_ids:
+                if not self._sample_exists(sample_id):
+                    raise ValueError('Sample ID %s does not exist.'
+                                     % sample_id)
+                if self._study_sample_are_associated(study_id, sample_id):
+                    raise ValueError('Study ID %s and sample ID %s are already'
+                                     ' associated.' % (study_id, sample_id))
+                sql = """INSERT INTO pm.study_sample (study_id, sample_id)
+                         VALUES (%s, %s)"""
+                TRN.add(sql, [study_id, sample_id])
+
+    def dissociate_samples_from_study(self, study_id, sample_ids):
```

Comment:
@antgonza The implementation is also among my concerns. Also, is it absolutely not possible that some samples are out there without an associated study? Could it happen that the wet lab people found that one sample is mistakenly assigned to a study. Then they have to unlink the study, but they cannot figure out immediately which study this sample actually belongs to. So they would rather leave this empty for the moment. Therefore, allowing it to be 0 gives people more flexibility. @josenavas If "person" and "date" of a plate can be empty (as you suggested), I think "study" of a sample can also be empty.